### PR TITLE
[NTGDI] Prevent NULL pointer dereference in CLIPPING_UpdateGCRegion

### PR DESCRIPTION
**Pull request by tkreuzer**
This can happen, when the paged pool is full. This will probably cause drawing issues, but at least we don't crash (in this place).

### CODE REVIEW

**Review by whindsaks — APPROVED** [review]
Should it print a debug message in the failure case or would it print too often when the system ends up in this state?

**Comment by tkreuzer** [comment]
> Should it print a debug message in the failure case or would it print too often when the system ends up in this state?

That is probably a good idea. Will add.

**Review by binarymaster — APPROVED** [review]

File: win32ss/gdi/ntgdi/cliprgn.c [hunk 1]
```diff
@@ -804,6 +811,13 @@ CLIPPING_UpdateGCRegion(PDC pDC)
         }
     }
 
+    if (!pDC->prgnRao)
```
Comment (HBelusca) [inline win32ss/gdi/ntgdi/cliprgn.c:814]: Why don't you move this warning (+ return) just after where the prgnRao is created? (i.e. line 785)
Or maybe, instead, move the prgnRao allocation down here?

**Review by HBelusca — APPROVED** [review]
OK for the NULL checks; I'm just wondering about what I mentioned in the comment.

File: win32ss/gdi/ntgdi/cliprgn.c (hunk 1 above)
Comment (tkreuzer) [reply win32ss/gdi/ntgdi/cliprgn.c:814]: To give the code above a chance to run.

Comment (HBelusca) [reply win32ss/gdi/ntgdi/cliprgn.c:814]: ok then what about moving the rao alloc down too? :thinking:

Comment (tkreuzer) [reply win32ss/gdi/ntgdi/cliprgn.c:814]: Sounds good, will do.

**Review by DarkFire01 — APPROVED** [review]

File: win32ss/gdi/ntgdi/cliprgn.c [hunk 2]
```diff
@@ -804,6 +806,17 @@ CLIPPING_UpdateGCRegion(PDC pDC)
...
     }
 
+    if (pDC->prgnRao)
+        REGION_Delete(pDC->prgnRao);
+
+    pDC->prgnRao = IntSysCreateRectpRgn(0,0,0,0);
+    if (!pDC->prgnRao)
+    {
+        /* Best we can do here. Better than crashing. */
+        ERR("Failed to allocate prgnRao! Expect drawing issues!\n");
```
Comment (JoachimHenze) [inline win32ss/gdi/ntgdi/cliprgn.c:816]:
```suggestion
        ERR("Failed to allocate prgnRao! Expect drawing issues\n");
```

File: win32ss/gdi/ntgdi/cliprgn.c [hunk 3]
```diff
@@ -776,16 +778,16 @@ CLIPPING_UpdateGCRegion(PDC pDC)
...
-
-    ASSERT(pDC->prgnRao);
-
     if (pDC->dclevel.prgnMeta || pDC->dclevel.prgnClip)
     {
         pDC->prgnAPI = IntSysCreateRectpRgn(0,0,0,0);
+        if (!pDC->prgnAPI)
+        {
+            /* Best we can do here. Better than crashing. */
+            ERR("Failed to allocate prgnAPI! Expect drawing issues!\n");
```
Comment (JoachimHenze) [inline win32ss/gdi/ntgdi/cliprgn.c:787]:
```suggestion
            ERR("Failed to allocate prgnAPI! Expect drawing issues\n");
```

Comment (tkreuzer) [reply win32ss/gdi/ntgdi/cliprgn.c:787]: `ERR("Failed to provide useful feedback! Stop wasting my time!!!!\n");`

**Pull request merged**